rat/compiler.py

```diff
@@ -1,488 +1,439 @@
-                        integer_shifts.append(folded_shift)
-                    except Exception as e:
-                        error_msg = f"Failed evaluating shift. Shift amount expressions must be an expression which can be evaluated at compile-time."
-                        raise CompileError(error_msg, shift_expr.range) from e
-
-        # If there is a non-zero shift, set pad needed to True
-        pad_needed = any(shift != 0 for shift in integer_shifts)
-
-        return tuple(integer_shifts), pad_needed
-
-    def build_variable_table(self):
-        """
-        Builds the "symbol table", which holds information for all variables in the model.
-        """
-        self.variable_table = VariableTable(self.data)
-
-        # Add entries to the symbol table for all data/params
-        for top_expr in self.expr_tree_list:
-            for symbol in top_expr:
+                        # It's okay if an argument_count == 0 something doesn't bind -- it (hopefully) will be provided by
+                        #  the primary variable instead
+                        return
+            else:
+                record = ConstantVariableRecord(node.name, argument_count)
+                try:
+                    record.bind(subscript_names, self.data)
+                except KeyError:
+                    record = SampledVariableRecord(node.name, argument_count)
+
+            self.variable_table[node.name] = record
+        else:
+            record = self.variable_table[node.name]
+            if isinstance(record, ConstantVariableRecord):
                 try:
-                    match symbol:
-                        case ast.Data():
-                            self.variable_table.insert(variable_name=symbol.get_key(), variable_type=VariableType.DATA)
-                        case ast.Param():
-                            self.variable_table.insert(variable_name=symbol.get_key(), variable_type=VariableType.PARAM)
+                    record.bind(subscript_names, self.data)
                 except KeyError as e:
-                    raise CompileError(str(e), symbol.range)
+                    if argument_count > 0:
+                        raise CompileError(str(e), node)
+                    else:
+                        # I think it's okay if an argument_count == 0 something doesn't bind here.
+                        return
 
-        # TODO: I'm not sure it's a good pattern to modify the AST in place
-        # Fold all shifts to constants
-        for top_expr in self.expr_tree_list:
-            for symbol in top_expr:
-                match symbol:
-                    case ast.Data() | ast.Param():
-                        if symbol.subscript is not None:
-                            integer_shifts, padding_needed = self._get_shifts_and_padding(symbol.subscript)
+        if node.arglist:
+            with self.in_control_flow.push(True):
```

Comment:
Hmm, I guess the funny thing here is that we never push false. It's you go into control flow and you're there. I think this could just be replaced by a default-to-false bool. Does that cover your Q?

Reply:
Yeah switching to just a bool seems better.

Reply:
https://github.com/bbbales2/regressions/blob/37b9d2f8c7c4d63a59db9df1079d8fa08018eb0d/rat/codegen_backends.py#L53

Well it does pop the `True` value after it's done though. I actually think we can keep it as-is.